docker/docker_client.go

```diff
@@ -214,17 +212,33 @@ func hasFile(files []os.FileInfo, name string) bool {
 
 // newDockerClient returns a new dockerClient instance for refHostname (a host a specified in the Docker image reference, not canonicalized to dockerRegistry)
 // “write” specifies whether the client will be used for "write" access (in particular passed to lookaside.go:toplevelFromSection)
-func newDockerClient(ctx *types.SystemContext, ref dockerReference, write bool, actions string) (*dockerClient, error) {
-	registry := reference.Domain(ref.ref)
+func newDockerClient(ctx *types.SystemContext, ref dockerReference, write bool, username, password, registry, actions string) (*dockerClient, error) {
+	tr := newTransport()
+	tr.TLSClientConfig = serverDefault()
+	client := &http.Client{Transport: tr}
+
+	if actions == "login" {
```

Comment:
This seems a bit surprising at a first glance, though I didn’t read the code enough to understand why it is necessary or whether there are alternatives yet.

Reply:
In the other path, `username` and `password` are supplied through `SystemContext`, and `registry` through `ref`. Having two sets of mutually exclusive parameters with the same semantics is confusing.

Perhaps `newDockerClient` should be split into three, an underlying constructor, a wrapper for the `types.Image*` uses, and a wrapper for `CheckAuth` (perhaps inlined into `CheckAuth`), with the wrappers extracting information from inputs and calling the underlying constructor.

Reply:
Right, I was about to ask what the best of way of handling this is. Could you explain a bit more about splitting it into 3 different parts.

Reply:
Something like
`newDockerClientWithDetails(ctx, hostname, username, password, …)` which does the TLS and cert setup and the like (`CheckAuth` will need to use the certificates as well), and
```go
func newDockerClient(ctx, ref, …) {
  hostname := reference.Domain(ref.ref)
  username, password, err := config.GetAuthentication(ctx, hostname 
  return newDockerClientWithDetails(ctx, hostname, username, password, …)
}
```

i.e. instead of having a `actions == "login"` conditional at runtime, separate the two paths by code structure. 

(The `newDockerClientWithDetails` name is really bad though, it says nothing about what the difference is. `newDockerClient`/`newDockerClientFromRef`?)

With this, there wouldn’t be a third part, `CheckAuth` would call `newDockerClientWithDetails` directly).

---

For the `CheckAuth` path, the caller might want to opt out of `configuredSignatureStorageBase`, again by moving it into the other wrapper. (Computing that wouldn’t hurt much, but it is unnecessary work for that purpose.)

---

I can’t see the `CheckAuth` path doing anything about the data in `authScope`, though; how does that work? Is it legal, and does it work, to say `repository::`?  If I am looking at docker/docker correctly (untested!), it is omitting the `scope` parameter entirely in that case; hopefully that’s good enough – but again, the two paths will need to supply different data to the created object.

Reply:
okay makes sense, thanks. Will get started on this.

Reply:
(Note to self: `authScope` still outstanding.)

Reply:
(The `authScope` part is still outstanding.)

Reply:
Isn't authScope needed only when you are doing a push or pull?

Reply:
`setupRequestAuth` uses it independent of what the operation is; i.e. even `CheckAuth` may end up using the value.

(I haven’t checked, maybe the `repository::` value, which the code seems to use in that case, is legal and OK to send; have you checked and is that intended, or is it just something that happens to work right now in testing?)